templates/rustdoc.hbs

```diff
@@ -16,19 +16,37 @@
 <script type="text/javascript" src="/menu.js?{{cratesfyi_version_safe}}"></script>
 </body>
 <script>
-    var doc_body = document.getElementById("rustdoc_body_wrapper");
-    if(window.location.hash) {
-        var notFirefox = typeof InstallTrigger === 'undefined';
-        if(notFirefox) {
-            var hash = window.location.hash;
-            window.location.hash = "";
-            setTimeout(function () {
-                window.location.hash = hash;
-                doc_body.focus();
-            }, 1);
+  // Reset the scroll offset on browsers that don't support
+  // scroll-padding-top (Desktop & Mobile Safari):
+  (function() {
+    const maybeFixupViewPortPosition = function() {
+      if (window.location.hash) {
+        const anchorElement = document.getElementById(window.location.hash.substr(1));
+        const navBarHeight = document.getElementsByClassName("nav-container-rustdoc")[0].offsetHeight;
+        if (anchorElement &&
+            anchorElement.getBoundingClientRect().top <= navBarHeight &&
+            Math.abs(anchorElement.getBoundingClientRect().top) >= 0) {
+          // It's just overlapped by the nav bar. This can't be a coincidence, scroll it into view:
+          requestAnimationFrame(function() {
+            scrollBy(0, -navBarHeight);
+          });
         }
-    } else {
-        doc_body.focus();
+      }
+    };
+    window.addEventListener("hashchange", maybeFixupViewPortPosition, false);
+    var usedBack = false;
+    if(window.performance) {
+      var navEntries = window.performance.getEntriesByType('navigation');
+      if(navEntries.length > 0 && navEntries[0].type === 'back_forward') {
+        usedBack = true;
+      } else if(window.performance.navigation
+                && window.performance.navigation.type == window.performance.navigation.TYPE_BACK_FORWARD) {
+        usedBack = true;
+      }
```

Comment:
I refactored it to removed the unnecessary `if` and make its purpose slightly clearer. I'm not sure about the indentation, I don't know what JS conventions are.

```suggestion
    if(window.performance) {
      var navEntries = window.performance.getEntriesByType('navigation');
      var usedBack = navEntries.length > 0 && navEntries[0].type === 'back_forward'
        || (window.performance.navigation
            && window.performance.navigation.type == window.performance.navigation.TYPE_BACK_FORWARD);
```